prboom2/src/MUSIC/libopenmptplayer.c

```diff
@@ -0,0 +1,195 @@
+/* Emacs style mode select   -*- C -*-
+ *-----------------------------------------------------------------------------
+ *  This program is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU General Public License
+ *  as published by the Free Software Foundation; either version 2
+ *  of the License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ *  02111-1307, USA.
+ *
+ * DESCRIPTION:
+ *  Tracker music player based on libopenmpt.
+ *---------------------------------------------------------------------
+ */
+
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "musicplayer.h"
+
+#ifndef HAVE_LIBOPENMPT
+#include <stddef.h>
+
+static const char *mpt_name (void)
+{
+  return "libopenmpt tracker player (DISABLED)";
+}
+
+
+static int mpt_init (int samplerate)
+{
+  return 0;
+}
+
+const music_player_t mpt_player =
+{
+  mpt_name,
+  mpt_init,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL
+};
+
+#else // HAVE_LIBOPENMPT
+
+#include <libopenmpt/libopenmpt.h>
+#include <string.h>
+#include <stdlib.h>
+#include <math.h>
+#include "lprintf.h"
+
+
+static int mpt_samplerate;
+static int mpt_paused;
+static int mpt_gain;
+static openmpt_module *mpt_cur_module;
+
+static const char *mpt_name (void)
+{
+  return "libopenmpt tracker player";
+}
+
+
+static int mpt_init (int samplerate)
+{
+  mpt_samplerate = samplerate;
+
+  return 1;
+}
+
+static void mpt_shutdown (void)
+{
+}
+
+static void mpt_setvolume (int v)
+{
+  int ret = 1;
+
+  // For 0% volume, log10f(0.f) yields -inf, which after converting to
+  // int yields INT_MIN gain.
+  mpt_gain = 2000.f * log10f(v / 15.f);
+
+  if (mpt_cur_module)
+    ret = openmpt_module_set_render_param(mpt_cur_module,
+                    OPENMPT_MODULE_RENDER_MASTERGAIN_MILLIBEL,
+                    mpt_gain);
+
+  if (!ret)
+    lprintf(LO_WARN, "mpt_setvolume: failed to set current module master gain\n");
+}
+
+static const void* mpt_registersong (const void *data, unsigned len)
+{
+  openmpt_module *mod = NULL;
+
+  mod = openmpt_module_create_from_memory2(data, len, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+  if (!mod)
+    lprintf(LO_WARN, "mpt_registersong: failed to open song\n");
```

Comment:
An effort was made to clean up terminal messages in DSDA-Doom compared to PrBoom+. We should set the `LO_DEBUG` level for this `lprintf`; otherwise, it will always print an error when a regular MIDI file is registered.

Another issue is that `openmpt_module_create_from_memory2` always prints `openmpt: openmpt_module_create_from_memory2: ERROR: error loading file` if non-tracker music is registered. This is not a big problem, but perhaps we could try defining [openmpt_error_func](https://lib.openmpt.org/doc/group__libopenmpt__c.html#gaa68155b220861de6218bc6b599071f21) or [openmpt_log_func](https://lib.openmpt.org/doc/group__libopenmpt__c.html#gaeec3c004981950af696bfd2688e250b2).

Reply:
Will change the log level. For libopenmpt messages just a custom `openmpt_log_func` is enough, the default error func just logs via whatever the log func is (and tries to store the message but we don't give it anywhere to do so). I've made the custom log func just print the message with the debug log level as well.